### ISSUE TITLE
Poor performance for looped sequence of many conditional `re.sub` calls

### ISSUE THREAD

**Issue by liblit** [microsoft/pyright#4588]
**Describe the bug**

Pyright 1.1.292 takes an exceptionally long time (or perhaps forever) to analyze a specific piece of Python code consisting of a `for` loop whose body is a long sequence of `if`-conditioned calls to `re.sub`.

**To Reproduce**

1. Save the code given below as `script.py` in an otherwise empty directory.

1. Run `pyright script.py`.

1. Wait.

**Expected behavior**

The `pyright` command should have completed within a second or two. Any errors reported are immaterial: the point is that the command should run and be done quickly.

**Screenshots or Code**

```python
import re


# noinspection PyUnboundLocalVariable,PyUnusedLocal
def process(items, flag):
    for item in items:
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
        if flag:
            text = re.sub("a", "b", text)
```

**Additional context**

Pyright 1.1.292 completes in 1.3 seconds if `text` is a function argument, as in:

```python
def process(items, flag, text):
```

The `for` loop is important.  Pyright 1.1.292 completes in 0.8 seconds if the `for` loop is removed, with all `if` statements lifted into the function body.

The `re.sub` calls are important.  Pyright 1.1.292 completes in 0.6 seconds if all calls to `re.sub` are replaced with calls to some unknown callee, such as `re_sub`.

Pyright 1.1.162 analyzes this file in 1.0 seconds. Pyright 1.1.163 has been running for more than one hour and is still not done.  So this problem seems to have arisen sometime between those two releases. The problem persists in Pyright 1.1.292, though. In particular, this problem was not fixed by #4519.

**Comment by erictraut** [comment]
I'm able to repro the problem as described, and I'll look into it.

Out of curiosity, how is the variable `text` initialized in this code? It doesn't appear to be given an initial value anywhere, so I'd expect it to be unbound and generate a runtime exception when this code is executed.

**Comment by liblit** [comment]
Thank you for your amazingly speedy response, @erictraut!

> Out of curiosity, how is the variable text initialized in this code? It doesn't appear to be given an initial value anywhere, so I'd expect it to be unbound and generate a runtime exception when this code is executed.

You're right that this code would fail at run time due to manipulating an uninitialized variable. As you might guess, the code given here is simplified and abstracted from the original. The code given here is goofy and useless, other than triggering the same bug as the original code (which I cannot share here).

I can tell you that in the original code, `text` appears to have been initialized earlier to a value read from a file using the open file's `readline()` method. I stripped that out from the code given here because it was not essential to reproduce the problem. However, if including the initialization(s) is essential to help you find a fix, then I'll see if I can produce a new problem-triggering code example that meets those criteria.

**Comment by erictraut** [comment]
I'm a bit at a loss of what to do with this. I'm going to keep the issue open and continue to think about it, but I wanted to give you a quick update on what I've found so far.

The problem is that this code requires 2^n separate code flow passes (where n is the number of if statements in the loop). It needs to consider every combination of "taken" and "not taken" for all of the `if` statements to evaluate the type of `text`. Pyright contains a bunch of optimizations to avoid this sort of exponential growth, but this particular code defeats all of these optimizations, and there's really no way to shortcut it in this case.

To make matters worse, `re.sub` is an overloaded method. Evaluating call expressions with overloads is at least 3x more expensive than non-overloads, so that's contributing to the cost here.

Pyright also contains some code to compute the cyclomatic complexity of a function prior to attempting to evaluate the types within the function. If the complexity exceeds a certain threshold, it doesn't make an attempt. In this case, the cyclomatic complexity isn't that huge, so the function falls well below the threshold.

Earlier versions of pyright contained some bugs that caused them to give up early when analyzing code like this. It was a bug fix in 1.1.163 that caused the evaluation time to increase.

The best answer in this case, unfortunately, may be for you to modify your code. I don't know what the original code looks like since you've provided a simplified version here. But perhaps there are ways to refactor the code using subroutines rather than a long list of `if` statements?

If it's not feasible to refactor your code, you could enclose the entire loop in an `if not TYPE_CHECKING:` block, which will cause pyright to not analyze it.

**Comment by erictraut** [comment]
I found the root cause of the problem and implemented a fix. This will be included in the next release.

**Comment by liblit** [comment]
Hooray! Thank you, @erictraut.

**Comment by erictraut** [comment]
This is included in pyright 1.1.296, which I just published. It will also be included in a future release of pylance.

**Comment by liblit** [comment]
Thank you!
